Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -806,116 +940,143 @@ static __forceinline u32 TranslatePageAddress(const u32 address, const XCheckTLB
+			// implemented this way for invalid BATs as well.
+			WARN_LOG(POWERPC, "Bad BAT setup: BPRN overlaps BL");
+		}
+		if (CountSetBits((u32)(batu.BL + 1)) != 1)
+		{
+			// With a valid BAT, the simplest way of masking is
+			// (input & ~BL_mask) for matching and (input & BL_mask) for
+			// translation. For now, assume it's implemented this way for
+			// invalid BATs as well.
+			WARN_LOG(POWERPC, "Bad BAT setup: invalid mask in BL");
+		}
+		for (u32 j = 0; j < (1 << 11); ++j)
 		{
-			// bat applies to this address
-			if (PowerPC::ppcState.spr[spr + i * 2] & batu)
+			// Enumerate all bit-patterns which fit within the given mask.
+			if ((j & batu.BL) == j)
 			{
-				// bat entry valid
-				u32 offset = BAT_EA_OFFSET(addr);
-				u32 page = BAT_EA_11(addr);
-				page &= ~bl17;
-				page |= BATL_BRPN(PowerPC::ppcState.spr[spr + 1 + i * 2]);
-				// fixme: check access rights
-				*result = page | offset;
-				return true;
+				// This bit is a little weird: if BRPN & j != 0, we end up with
+				// a strange mapping. Need to check on hardware.
+				u32 address = (batl.BRPN | j) << 17;
+				// (BEPI | j) == (BEPI & ~BL) | (j & BL).
+				bat_table[batu.BEPI | j] = address | 0x1;
 			}
 		}
 	}
-	return false;
 }
 
-// Block Address Translation
-static u32 TranslateBlockAddress(const u32 address, const XCheckTLBFlag flag)
+static void UpdateFakeMMUDBat(u32 start_addr)
 {
-	u32 result = 0;
-	UReg_MSR& m_MSR = ((UReg_MSR&)PowerPC::ppcState.msr);
-	u32 batu = (m_MSR.PR ? BATU_Vp : BATU_Vs);
-
-	// Check for enhanced mode (secondary BAT enable) using 8 BATs
-	bool enhanced_bats = SConfig::GetInstance().m_LocalCoreStartupParameter.bWii && HID4.SBE;
-
-	if (flag != FLAG_OPCODE)
+	for (unsigned i = 0; i < (0x10000000 >> 17); ++i)
 	{
-		if (!CheckAddrBats(address, &result, batu, SPR_DBAT0U) && enhanced_bats)
-			CheckAddrBats(address, &result, batu, SPR_DBAT4U);
+		// Map from 0x4XXXXXXX or 0x7XXXXXXX to the range
+		// [0x7E000000,0x80000000).
+		u32 e_address = i + (start_addr >> 17);
+		u32 p_address = 0x7E000001 | ((i << 17) & Memory::FAKEVMEM_MASK);
+		dbat_table[e_address] = p_address;
 	}
-	else
+}
+
+void DBATUpdated()
+{
+	memset(dbat_table, 0, sizeof(dbat_table));
+	UpdateBATs(dbat_table, SPR_DBAT0U);
+	bool extended_bats = SConfig::GetInstance().m_LocalCoreStartupParameter.bWii && HID4.SBE;
+	if (extended_bats)
+		UpdateBATs(dbat_table, SPR_DBAT4U);
+	if (Memory::bFakeVMEM)
 	{
-		if (!CheckAddrBats(address, &result, batu, SPR_IBAT0U) && enhanced_bats)
-			CheckAddrBats(address, &result, batu, SPR_IBAT4U);
+		// In Fake-MMU mode, insert some extra entries into the BAT tables.
+		UpdateFakeMMUDBat(0x40000000);
+		UpdateFakeMMUDBat(0x70000000);
 	}
-	return result;
+	Memory::UpdateLogicalMemory(dbat_table);
+	JitInterface::ClearSafe();
+}
+
+void IBATUpdated()
+{
+	memset(ibat_table, 0, sizeof(ibat_table));
+	UpdateBATs(ibat_table, SPR_IBAT0U);
+	bool extended_bats = SConfig::GetInstance().m_LocalCoreStartupParameter.bWii && HID4.SBE;
```

Comment:
Yes.  (Granted, I should probably set HID4 in our BS2 emulation for Wii disks and homebrew, just to be consistent.)